### PR TITLE
Allow the observed node to be specified and the selector scan search pattern to be configured.

### PR DESCRIPTION
**Pull request by dbezborodovrp**
Support scanning of only addedNodes in the mutation instead of scanning the entire document for matching selectors.

### CODE REVIEW

**Comment by dbezborodovrp** [comment]
This is for #17

**Comment by pie6k** [comment]
Maybe it'd be better idea to add option like `rootNode` (that obviously can be `document`)?

**Comment by dbezborodovrp** [comment]
Hi @pie6k, yes, we could probably improve performance by passing in a root node to observe. However, I think this solution is valid also. They solve two different separate performance issues: 1) What object to observe for mutations, and 2) Where to scan for selectors when a mutation is observed. This change solves the latter by making it possible to only scan the added nodes in the mutation (instead of the entire document.)

If we only implemented passing in a rootNode, you would still need to scan the entire root node for matching selectors without this change.

This pull request allows, for example:

  1) `rootNode` is observed (usually it would be `document.documentElement`)
  2) There is a `<table>` in the body.
  3) An initializer with a selector `tr.exampleClass` is registered.
  4) A row `<tr class="exampleClass">` is added to the table.
  5) At this point, a mutation event is triggered, which will have:
     - `target: document.documentElement`
     - `addedNodes: [tr]`

----

With this change, if `scanDocument` is false, then the mutation observer will only search for matching selectors within `addedNodes`.

If we wanted to implement your suggestion alongside this, we could improve this even further. For example:

1) There is a `<table>` in the body.
2) An initializer with a selector `tr.exampleClass` is registered, observing the `<table>` element instead of the entire document. (`rootNode = table`)
3) A row `<tr class="exampleClass">` is added to the table.
4) At this point, a mutation event is triggered, which will have:
   - **`target: table`**
   - `addedNodes: [tr]`

So, there is potential to improve two things: 1) which element mutations are observed on, and 2) how to match selectors when the mutation event is captured.

I will see if I can improve this pull request further to make both features possible.

File: jquery.initialize.js [hunk 1]
```diff
@@ -34,26 +38,51 @@
...
+
+                // Scan the entire document.
+                $(msobservers[j].selector)
+                    .each(msobservers[j].callback);
+            } else {
+
+                // For each mutation.
+                for (var m = 0; m < mutations.length; m++) {
+                    // If mutation type is observed.
+                    if ($.inArray(mutations[m].type, mtypes) != -1) {
```
Comment (pie6k) [inline jquery.initialize.js]: For code I'd consider negative `ifs` eg
```
if (!$.inArray(mutations[m].type, mtypes) != -1) {
  continue;
}
// rest of the code. HEY! we caved some tabs in indent!
```

Comment (dbezborodovrp) [reply jquery.initialize.js]: I will update accordingly as I implement your other concerns.

**Comment by dbezborodovrp** [comment]
I might also need to rename the `scanDocument` option to something more meaningful.

File: jquery.initialize.js (hunk 1 above)
Comment (pie6k) [reply jquery.initialize.js]: Don't forget to add yourself to contributors as well as some note in the README

**Comment by dbezborodovrp** [comment]
So, `rootNode` would be called `target`. I'll use that terminology from now on.

```
MutationObserver.observe(Node target, MutationObserverInit options)
```

**Comment by dbezborodovrp** [comment]
Oops, not entirely correct. If the mutation is an `attributes` mutation, then `target` would be the node that the attribute was changed on.

**Comment by dbezborodovrp** [comment]
I have updated the pull request to allow the "root node" to be specified and will now test it in my company's application, to see how it affects performance. Then, I might consider renaming the options after that. See also, test2.html.

**Comment by dbezborodovrp** [comment]
@pie6k I have this working now and is ready for testing.

I had to use JavaScript that may break older browsers (that is, [Array.forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach), [Element.querySelectorAll()](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelectorAll) and [Element.matches()](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches))

There options are now:
```
options.target: Document | Element                        The element to observe mutations on.
                                                          Defaults to document.documentElement

options.scanMode: 'target' | 'descendants' | 'exact'      Selector search strategy.
```

Explanation of the different scan modes are:

 * target: Query the target element (by default, the entire document) for elements that match the given selector.
 * descendants: Query the node that was mutated or added to the object model for elements that match the given selector (including the muted node and child nodes.)
 * exact: Check if the node that was mutated or added to the object model matches the selector. This does not include child nodes.

This should provide a large degree of control to the developer. Unfortunately, the best way to understand these scanModes is by using them. **See the example** in `test2.html`, and try using different scanModes.

**Comment by dbezborodovrp** [comment]
After you are happy with this, I can document it in the README before merging.

**Comment by pie6k** [comment]
Seems good. I think it'd be good to provide links to proper demos in README

**Comment by pie6k** [comment]
Yup, I think GH is totally fine. Thanks for your help @dbezborodovrp

**Comment by dbezborodovrp** [comment]
I opened a separate issue for documentation, #20.

**Comment by pie6k** [comment]
Cool, merged 🎉

**Comment by dbezborodovrp** [comment]
Cool. I think I could have actually simplified this. I might open a second PR if you can hold off tagging a new version number.

**Comment by dbezborodovrp** [comment]
See #21 for simplification of this feature.

**Pull request merged**